# Polynomial Regression is a form of linear regression in which the relationship between the independent variable x and dependent variable y is modeled as an nth degree polynomial. Polynomial regression fits a nonlinear relationship between the value of x and the corresponding conditional mean of y, denoted E(y |x).

# For degree=2, the linear modelis built. The mean squared error is calculated and r squared is found for training and testing.

# Import Libraries

In [1]:
#importing numpy and pandas, seaborn

import numpy as np #linear algebra
import pandas as pd #datapreprocessing, CSV file I/O
import seaborn as sns #for plotting graphs
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import PolynomialFeatures
from sklearn import metrics
from mpl_toolkits.mplot3d import Axes3D
%matplotlib inline

# Read File

In [2]:
house = pd.read_csv("C:/Users\Shermaine/Documents/Python ML II (Cher Wah)/kc_house_data.csv")

df = pd.DataFrame(house)

# Train_Test_Split= 80/20

In [3]:
train_data,test_data=train_test_split(df,train_size=0.8,random_state=3)

#same as in multiple linear regression
#x = df[['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot','floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement','yr_built', 'yr_renovated', 'zipcode', 'lat','sqft_living15','sqft_lot15']]
#features1= ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot','floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement','yr_built', 'yr_renovated', 'zipcode', 'lat','sqft_living15','sqft_lot15']

features= df.drop(['id','date','price', 'floors', 'sqft_lot'],axis=1)
print(features)

features1=x = ['bedrooms', 'bathrooms', 'sqft_living', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement','yr_built', 'yr_renovated', 'zipcode', 'lat','sqft_lot15']

       bedrooms  bathrooms  sqft_living  waterfront  view  condition  grade  \
0             3       1.00         1180           0     0          3      7   
1             3       2.25         2570           0     0          3      7   
2             2       1.00          770           0     0          3      6   
3             4       3.00         1960           0     0          5      7   
4             3       2.00         1680           0     0          3      8   
...         ...        ...          ...         ...   ...        ...    ...   
21608         3       2.50         1530           0     0          3      8   
21609         4       2.50         2310           0     0          3      8   
21610         2       0.75         1020           0     0          3      7   
21611         3       2.50         1600           0     0          3      8   
21612         2       0.75         1020           0     0          3      7   

       sqft_above  sqft_basement  yr_built  yr_reno

# Train and Time Taken, Degree= 2

In [4]:
import time

polyfeat=PolynomialFeatures(degree=2)

start_time=time.time()

xtrain_poly=polyfeat.fit_transform(train_data[features1])

xtest_poly=polyfeat.fit_transform(test_data[features1])

poly=linear_model.LinearRegression()
poly.fit(xtrain_poly,train_data['price'])

end_time=time.time()

print("---%s seconds ---" % (end_time - start_time))

polypred=poly.predict(xtest_poly)

---0.11967992782592773 seconds ---


# Print Results

In [5]:
print('PolynomialRegressionRebuild')
mean_squared_error=metrics.mean_squared_error(test_data['price'],polypred)
print('Mean Squared Error (MSE) ', round(np.sqrt(mean_squared_error), 2))
print('R-squared (training) ', round(poly.score(xtrain_poly, train_data['price']), 3))
print('R-squared (testing) ', round(poly.score(xtest_poly, test_data['price']), 3))

PolynomialRegressionRebuild
Mean Squared Error (MSE)  162705.19
R-squared (training)  0.799
R-squared (testing)  0.794


# R-squared testing: 0.794

# Rebuild Model for degree=3.The mean squared error is calculated and r squared is found for training and testing.

# Train and Time Taken, Degree= 3

In [6]:
import time

polyfeat=PolynomialFeatures(degree=3)

start_time=time.time()

xtrain_poly=polyfeat.fit_transform(train_data[features1])

xtest_poly=polyfeat.fit_transform(test_data[features1])

poly=linear_model.LinearRegression()
poly.fit(xtrain_poly,train_data['price'])

end_time=time.time()

print("---%s seconds ---" % (end_time - start_time))

polypred=poly.predict(xtest_poly)

---1.0571742057800293 seconds ---


# Print Results

In [7]:
print('PolynomialRegressionRebuild')
mean_squared_error=metrics.mean_squared_error(test_data['price'],polypred)
print('Mean Squared Error (MSE) ', round(np.sqrt(mean_squared_error), 2))
print('R-squared (training) ', round(poly.score(xtrain_poly, train_data['price']), 3))
print('R-squared (testing) ', round(poly.score(xtest_poly, test_data['price']), 3))

PolynomialRegressionRebuild
Mean Squared Error (MSE)  206770.69
R-squared (training)  0.826
R-squared (testing)  0.667


# R-squared testing: 0.667

# Observation
PolynomialRegression gives us R-squared (testing) score of 0.79. From above reports, we can conclude that Polynomial regression for degree=2, is best solution out of multiple linear regression and polynomial regression.